net: if: Extend the usage of rejoining the multicast groups

markaj-nordic opened the pull request:
It may happen that the interface is up but not yet running when we issue the rejoin_ipv6_mcast_groups(). This can be fixed by calling this function again right after the iface is set to 'running' state in the notify_iface_up handler.

pdgendt commented on subsys/net/ip/net_if.c:
```diff
@@ -4636,6 +4636,13 @@ bool net_if_is_offloaded(struct net_if *iface)
 
+static void rejoin_multicast_groups(struct net_if *iface)
+{
+#if defined(CONFIG_NET_NATIVE_IPV6)
+	rejoin_ipv6_mcast_groups(iface);
+#endif
```
Since you're touching this:

```suggestion
#if defined(CONFIG_NET_NATIVE_IPV6)
	rejoin_ipv6_mcast_groups(iface);
#else
	ARG_UNUSED(iface);
#endif
```

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -4654,6 +4661,7 @@ static void notify_iface_up(struct net_if *iface)
 	}
 
 	net_if_flag_set(iface, NET_IF_RUNNING);
+	rejoin_multicast_groups(iface);
```
Should we actually move the
```
	if (!net_if_is_offloaded(iface)) {
		/* Make sure that we update the IPv6 addresses and join the
		 * multicast groups.
		 */
		rejoin_multicast_groups(iface);
		net_if_start_dad(iface);
	}
```
from `net_if_up()` to here. There is the point-to-point check few lines below where these two statements could be placed. I am not sure we need to do the rejoining in two places.

markaj-nordic replied:
Ok, thanks for the hints. Now the rejoin_multicast is called only if the operational state is NET_IF_OPER_UP:

`	if (net_if_oper_state(iface) == NET_IF_OPER_UP) {
		if (prev_state != NET_IF_OPER_UP) {
			notify_iface_up(iface);
		}`
		
I presume that's ok?

markaj-nordic committed:
net: if: Extend the usage of rejoining the multicast groups

It may happen that the interface is up but not yet running
when we issue the rejoin_ipv6_mcast_groups(). This can be
fixed by calling this function again right after the iface
is set to 'running' state in the notify_iface_up handler.

Signed-off-by: Marcin Kajor <marcin.kajor@nordicsemi.no>

rlubos commented on subsys/net/ip/net_if.c:
```diff
@@ -4664,6 +4676,12 @@ static void notify_iface_up(struct net_if *iface)
+		 * multicast groups.
+		 */
+		rejoin_multicast_groups(iface);
+		net_if_start_dad(iface);
```
`iface_ipv6_start()` already covers DAD triggering (and checks for NET_IF_IPV6_NO_ND flag, which is important for OT case). I think it'd be enough to just move `rejoin_multicast_groups()` before we call that.

markaj-nordic replied:
Ok, I see. Makes sense.

rlubos commented:
@pdgendt Can you revisit? Thanks.
